src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,355 @@
+
+	private final KafkaAckCallbackFactory<K, V> ackCallbackFactory;
+
+	private final String[] topics;
+
+	private String groupId;
+
+	private String clientId = "message.source";
+
+	private long pollTimeout = DEFAULT_POLL_TIMEOUT;
+
+	private RecordMessageConverter messageConverter = new MessagingMessageConverter();
+
+	private Type payloadType;
+
+	private volatile Consumer<K, V> consumer;
+
+	private volatile Iterator<ConsumerRecord<K, V>> recordIterator;
+
+	private volatile Collection<TopicPartition> partitions;
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory, String... topics) {
+		this(consumerFactory, new KafkaAckCallbackFactory<K, V>(), topics);
+	}
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory,
+			KafkaAckCallbackFactory<K, V> ackCallbackFactory, String... topics) {
+		Assert.notNull(consumerFactory, "'consumerFactory' must not be null");
+		Assert.notNull(ackCallbackFactory, "'ackCallbackFactory' must not be null");
+		this.consumerFactory = consumerFactory;
+		this.ackCallbackFactory = ackCallbackFactory;
+		this.topics = topics;
+	}
+
+	protected String getGroupId() {
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the messsage converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized AbstractIntegrationMessageBuilder<Object> doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		if (this.recordIterator != null && this.recordIterator.hasNext()) {
+			record = this.recordIterator.next();
+		}
+		else {
+			synchronized (this.consumer) {
```

Comment:
![image](https://user-images.githubusercontent.com/1071590/34628556-bb7537e8-f232-11e7-9f4f-ae79bcb1aa16.png)


I suggest to really have some `private final Object consumerMonitor = new Object();` instead.
Semantically we will have the same synchronization logic, but will definitely do it against a shared singleton object.